federation/schema.go

```diff
@@ -0,0 +1,161 @@
+package federation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	_ "embed"
+	"errors"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	_ "github.com/ipld/go-ipld-prime/codec/dagcbor"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/record"
+	"github.com/multiformats/go-multicodec"
+)
+
+const (
+	stateHeadRecordDomain = "indexer"
+	stateHeadRecordCodec  = "/indexer/fed/stateHeadSignature"
+)
+
+var (
+	Prototypes struct {
+		Head                     schema.TypedPrototype
+		Snapshot                 schema.TypedPrototype
+		ProvidersIngestStatusMap schema.TypedPrototype
+		IngestStatus             schema.TypedPrototype
+
+		link ipld.LinkPrototype
+	}
+
+	//go:embed schema.ipldsch
+	schemaBytes []byte
+
+	_                         record.Record = (*Head)(nil)
+	stateHeadRecordCodecBytes               = []byte(stateHeadRecordCodec)
+)
+
+type (
+	Head struct {
+		Head      ipld.Link
+		Topic     *string
+		PublicKey []byte
+		Signature []byte
+	}
+	Snapshot struct {
+		Epoch       uint64
+		VectorClock uint64
+		Providers   ProvidersIngestStatusMap
+		Previous    ipld.Link
```

Comment:
> A previous snapshot does not seem necessary for reconciliation.

Right. 

The rationale for having a chain of snapshot is to provide opportunity for reputation measurements in the future since IPNI will need some sort of reputation system atop a permissionless federation setup. 

Side note: It might be worth bringing up this question on the spec PR as it relates more to protocol design itself than the implementation. I think it is worth writing a rationale for its existence in the spec at least, if we reach the consensus to keep it.